rust_src/src/editfns.rs

```diff
@@ -936,4 +941,84 @@ pub fn buffer_string() -> LispObject {
     unsafe { make_buffer_string_both(begv, begv_byte, zv, zv_byte, true) }
 }
 
+// Save current buffer state for `save-excursion' special form.
+// We (ab)use Lisp_Misc_Save_Value to allow explicit free and so
+// offload some work from GC.
+pub extern "C" fn save_excursion_save() -> LispObject {
+    let window = selected_window().as_window_or_error();
+
+    unsafe {
+        make_save_obj_obj_obj_obj(
+            point_marker(),
+            Qnil,
+            // Selected window if current buffer is shown in it, nil otherwise.
+            if window.contents.eq(current_buffer()) {
+                window.into()
+            } else {
+                Qnil
+            },
+            Qnil,
+        )
+    }
+}
+
+/// Save point, and current buffer; execute BODY; restore those things.
+/// Executes BODY just like `progn'.
+/// The values of point and the current buffer are restored
+/// even in case of abnormal exit (throw or error).
+///
+/// If you only want to save the current buffer but not point,
+/// then just use `save-current-buffer', or even `with-current-buffer'.
+///
+/// Before Emacs 25.1, `save-excursion' used to save the mark state.
+/// To save the marker state as well as the point and buffer, use
+/// `save-mark-and-excursion'.
+///
+/// usage: (save-excursion &rest BODY)
+#[lisp_fn(unevalled = "true")]
+pub fn save_excursion(args: LispObject) -> LispObject {
+    let count = c_specpdl_index();
+
+    unsafe { record_unwind_protect(Some(save_excursion_restore), save_excursion_save()) };
+
+    unsafe { unbind_to(count, progn(args)) }
+}
+
+/// Record which buffer is current; execute BODY; make that buffer current.
+/// BODY is executed just like `progn'.
+/// usage: (save-current-buffer &rest BODY)
+#[lisp_fn(unevalled = "true")]
+pub fn save_current_buffer(args: LispObject) -> LispObject {
+    let count = c_specpdl_index();
+
+    unsafe { record_unwind_current_buffer() };
+
+    unsafe { unbind_to(count, progn(args)) }
+}
+
+/// Execute BODY, saving and restoring current buffer's restrictions.
+/// The buffer's restrictions make parts of the beginning and end invisible.
+/// \(They are set up with `narrow-to-region' and eliminated with `widen'.)
```

Comment:
Yeah, that's in the original. I'm not sure if it has any significance. It doesn't show up in the Help display.

Reply:
That's for escaping . We don't need it in the rust comments.

Reply:
FWIW it's because some elisp tools scan backwards for lines starting with a parenthesis. The C code has to avoid starting a line in a multiline docstring with a `(`. I forget which tool it was, it might have been `scan-sexps`.